lib/queue.js

```diff
@@ -670,7 +670,12 @@ Queue.prototype.setHandler = function(name, handler) {
     } else {
       this.handlers[name] = function() {
         try {
-          return Promise.resolve(handler.apply(null, arguments));
+          const maybeThenable = handler.apply(null, arguments)
```

Comment:
Why is this change needed? According to the Promise specification, resolving a promise just chains that promise to the new promise, so there should not be any need to differentiate the two. This works the same as the "await" keyword, you can apply it to a value or to a promise, the result will always be the resolved value.

Reply:
This is needed because, as you point out, native Promises do not support cancellation. That makes this a bit tricky. In the original code, if `handler` returns a Bluebird promise with a `cancel()` method, `Promise.resolve()` will return a native Promise that resolves to whatever the Bluebird promise did - but since it is now a native Promise and those don't support cancellation, the `cancel()` method doesn't get carried along and is lost, along with any other features that Bluebird promises have.

The reason it works for native Promises with a cancel method added is that in the special case where you call `Promise.resolve()` with a native Promise, instead of creating a new native Promise that resolves to what that promise resolves to, Node simply returns the original promise, so the added-on cancel method remains intact.

This change makes it so we do that for any promise-like object (that is, an object with a `then` method, hence "thenable", which I've seen promise-like objects referred to elsewhere). If the return value from the handler looks like a promise, then we return that value without wrapping it in a native promise, which preserves any `cancel()` methods. If it doesn't look like a promise, we still wrap it in a `Promise.resolve()` as before.

I can add a comment here with a brief note about why we're doing this manually, which might be helpful.

Reply:
That got a bit long, so I'll try to sum up briefly: resolving a promise does chain that promise to a new promise, but it does so by creating a new promise that resolves to the same value as the old promise - so while the resolved value gets carried over, anything that is not supported in native promises (such as cancellation) does not.

Reply:
I understand. I think the best would be to refactor this as a special promiseResolve function that works the same for both native and non-native promises, then it will be easier to understand its intent as well.